### DISCUSSION TITLE
An attempt to integrate with io-uring

### DISCUSSION THREAD

**Discussion by hidva** [memc-rs/memc-rs#7] [Show and tell]
Recently, I was trying to integrate io_uring into Tokio, and [KuiBaDB/kbio](https://hidva.com/g?u=https://github.com/KuiBaDB/kbio) works now. After I saw your amazing project, I also tried to use kbio to replace tokio in memc, and [it](https://hidva.com/g?u=https://github.com/hidva/memc-rs/commit/091182bdc6a3e87280f4b9925021e450dc23957d) works too. And the benchmark shows a slight improvement in performance.

```bash
memtier_benchmark -p 11211 -P memcache_binary --json-out-file=kbio.benchmark.json --hdr-file-prefix=kbio.hdr --out-file=kbio.out --print-percentiles='50,99,99.9,99.99' --test-time=30  -t 16 -c 64
```

```bash
# kbio
[RUN #1 60%,  18 secs] 16 threads: 30122029 ops, 1707360 (avg: 1673177) ops/sec, 69.26MB/sec (avg: 67.85MB/sec),  0.60 (avg:  0.61) msec lat
[RUN #1 63%,  19 secs] 16 threads: 31803824 ops, 1681499 (avg: 1673615) ops/sec, 68.21MB/sec (avg: 67.87MB/sec),  0.61 (avg:  0.61) msec lat
[RUN #1 67%,  20 secs] 16 threads: 33498826 ops, 1694805 (avg: 1674674) ops/sec, 68.72MB/sec (avg: 67.92MB/sec),  0.60 (avg:  0.61) msec lat


# tokio
[RUN #1 60%,  18 secs] 16 threads: 29330563 ops, 1633659 (avg: 1629239) ops/sec, 66.26MB/sec (avg: 66.07MB/sec),  0.63 (avg:  0.63) msec lat
[RUN #1 63%,  19 secs] 16 threads: 30916355 ops, 1585547 (avg: 1626939) ops/sec, 64.31MB/sec (avg: 65.98MB/sec),  0.65 (avg:  0.63) msec lat
[RUN #1 67%,  20 secs] 16 threads: 32445441 ops, 1528823 (avg: 1622033) ops/sec, 62.01MB/sec (avg: 65.78MB/sec),  0.67 (avg:  0.63) msec lat
```

HAHAHA

**Comment by DariuszOstolski** [comment]
I was also thinking about integrating it in memc. I've noticed that tokio runtime seems to be little less performant than memcached one. AFAIR memcached is using scatter/gather io to limit syscalls so I'd be very happy to merge pull request for that integration.

Right now I'm working on LRU policy which is quite difficult to get it right in Rust.

BR,
Dariusz

**Comment by hidva** [comment]
At present, KuiBaDB/kbio is still a very early prototype. Its final effect is that it is fully compatible with Tokio on interfaces other than File IO, and the performance data here is also worth further understanding.


BR,
盏一

----------




--------------原始邮件--------------
发件人："Dariusz&nbsp;Ostolski ***@***.***&gt;;
发送时间：2021年12月28日(星期二) 凌晨5:22
收件人："memc-rs/memc-rs" ***@***.***&gt;;
抄送："盏一 ***@***.***&gt;;"Author ***@***.***&gt;;
主题：Re: [memc-rs/memc-rs] An attempt to integrate with io-uring (Discussion #7)
-----------------------------------

 

 
I was also thinking about integrating it in memc. I've noticed that tokio runtime seems to be little less performant than memcached one. AFAIR memcached is using scatter/gather io to limit syscalls so I'd be very happy to merge pull request for that integration.
 
Right now I'm working on LRU policy which is quite difficult to get it right in Rust.
 
BR,
 Dariusz
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by DariuszOstolski** [reply]
Hi,

Cannot wait until its compatible with Tokio, just let me know when I can give it a try.

BR,
Dariusz

**Reply by hidva** [reply]
You can already try it with the patch [54 additions and 98 deletions.](https://hidva.com/g?u=https://github.com/hidva/memc-rs/commit/ccf9a16f8d477fd8b80f3f6cd6223fc8f47801d4). However, the `time` feature is not currently supported, so I removed the expiration feature in the patch.

Updated benchmark:

```
# tokio
(avg: 76.85MB/sec), (avg:  2.16) msec

# kbio
82.15MB/sec 2.02 msec.
```

```
# client, A.B.173.2X, aarch64 32cpu aliyun ecs.c6r.8xlarge
memtier_benchmark -s A.B.173.2Y -p 11211 -P memcache_binary --json-out-file=kbio.benchmark.json --hdr-file-prefix=kbio.hdr --out-file=kbio.out --print-percentiles='50,99,99.9,99.99' --test-time=30  -t 32 -c 128

# server, A.B.173.2Y, x86_64 32 cpu aliyun ecs.c7.8xlarge.
./target/release/memcrsd -r 16 -c 8192 -b 8192 -l 0.0.0.0
```

It costs twice as much CPU utilization, but only 6% improvement. It's sad and there's still a long way to go.

BR,
盏一

**Reply by hidva** [reply]
Currently, the memc uses Tokio current thread runtime, which has no work-stealing capability. I tested the memc under Tokio Multi-Thread Scheduler, and the result is `19.60MB/sec, 8.47msec`. So if you want to use io_uring:

- If you do not need the work-stealing capability, [bytedance/monoio](https://hidva.com/g?u=https://github.com/bytedance/monoio) may be your choice.
- And if you do need the work-stealing, [KuiBaDB/kbio](https://hidva.com/g?u=https://github.com/KuiBaDB/kbio) may be your choice.

And Our experience using C++20 coroutines in [Hologres](https://hidva.com/g?u=https://www.aliyun.com/product/bigdata/hologram) tells us that work-stealing is very necessary

Happy New Year,
盏一

**Reply by DariuszOstolski** [reply]
Happy New Year,

Thank you a lot for your feedback. I'm little bit worried about performance regression when using Tokio with work stealing scheduler(which was my first choice but after comparing benchmarks with memcached I decided to go with current thread runtime). I'd like try kbio and monoio and compare them. Do you have benchmarks that would show necessity of work stealing scheduler?

BR,
Dariusz

**Reply by hidva** [reply]
The key to whether you need work-stealing is whether there may be an unbalanced load in the production environment. I don't know Memcached very well. It seems that all its operations are O(1), which means that the frequency of uneven load should be very low. Maybe you should try [bytedance/monoio](https://github.com/bytedance/monoio) first. 

memtier_benchmark will always generate a balanced load, work-stealing has no meaning for this=. =
